### PR TITLE
Fix #1491 Adding option for generation of qthelp ID's without keywords

### LINKED ISSUE #1491: QtHelp should generate identifiers for keywords in qhp files.
This is necessary for the http://doc-snapshot.qt-project.org/qt5-5.3/qhelpenginecore.html#linksForIdentifier API to work. 

Qt-generated qhp files include a id attribute in the keyword element. In CMake, we would need to have 

```
id="${domain_object_name}/${name}"
```

such as

```
id="command/add_library"
id="prop_tgt/COMPILE_DEFINITIONS"
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1491
- Originally reported by: Anonymous
- Originally created at: 2014-06-17T11:06:14.433

### PR DESCRIPTION
**Pull request by kreuzberger**
Adding option to generate lean IDs for qthelp builder

### Feature or Bugfix
Bugfix

### Purpose
Create keyword independent ids in qthelp builder for refering language independent links

### Relates
https://github.com/sphinx-doc/sphinx/issues/1491

### CODE REVIEW

**Comment by kreuzberger** [comment]
The last commit failed on AppVeyor due to an upload error. How to retrigger build without commiting?

File: tests/test_build_qthelp_leanid.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import pytest
+
+
+@pytest.mark.sphinx('qthelp', testroot='index')
+def test_qthelp_namespace(app, status, warning):
+    # default namespace
+    app.builder.build_all()
+    qhp = (app.outdir / 'Python.qhp').text()
+    assert '<keyword name="features (ID_FEATURE)" id="ID_FEATURE.features"' in qhp
```
Comment (tk0miya) [inline tests/test_build_qthelp_leanid.py:22]: Do you know what purpose this id parameter is used? (I've never used qthelp...)
I'd like to know why have we assigned IDs manually for the keyword tag.

And I also want to know why `qthelp_id_lean` option is needed. Is there any reason to choose IDs lean or not?

Comment (kreuzberger) [reply tests/test_build_qthelp_leanid.py:22]: My purpose is to open Context Specific Online Help from my application from the code. This supports en, de, fr Online Help depending on the language settings
There are too possibilities: Too look for keywords, they are normaly generated from Document filenames, and keywords from index. But they are are language specific.

In my german document e.g. i would add:

.. index:: Manuelle Suche (ID_SEARCH_MANUALLY)

in my englisch document e.g. i would add:
.. index:: Search Manually (ID_SEARCH_MANUALLY)

And in my code i can then add

So i have too options: 
in my code a can call linksForKeyword (tr("Search Manually")) and look for keywords with prior tr translation from qt. The main disadvantage is that you have too keep this in mind if you change the indexes.
Second is to use the ID for index entries.
This id should be a unique and language independent identifier.
So you can manually generate index entries in 3 different language rst documents and then use
linksForIdentifier("ID_SEARCH_MANUALLY") in the code

In the current implementation of sphinx everything gets now mixed. So the id contains the keyword and i have now language dependencies again. So the idea is to separate this clearly.

Cause my implementation breaks current behaviour this patch would make current implementations to fail. So i think its better to use an option for this and make the old behaviour deprecated to give people a chance to change implementations.

**Comment by kreuzberger** [comment]
Hi! Still missing this feature. Whats blocking this issue to be integrated into master? Can i help / support?

**Comment by tk0miya** [comment]
I'm sorry for not responding to your message...

If my understanding right, you're trying to change the meaning of `index` directive. At present, parenthesis is not special meaning. Simply, it is a part of keyword.
```
.. index:: features (ID_FEATURE)
```

With this change, it is considered as an ID for the keyword, right? I think it introduced new semantics only for qthelp. I feel that is too much.

**Comment by kreuzberger** [comment]
Hmm, you seem to be right, thought it has a special meaning due to different keyword and id generation.
That does not solve my problem, but i agree with you, no special meaning should be introduced in a builder.
My origin problem is that i want to generate language independent references to index entries, but this seems not to be possible at all.

**Comment by tk0miya** [comment]
How about making own custom qthelp builder? You can do it with subclassing qthelp builder and install it via `app.add_builder()` API.
Anyway, it is not good to merge this into Sphinx core to me.
Thanks,

**Comment by kreuzberger** [comment]
Yes, understand, you should not merge it.
I must rethink it. The main purpose still exists, when using sphinx-doc for my online help. In my c++ code i do not want to maintain language dependent references, just want to use a language independent identifier.
Most references i know require knowledge of the document structure, so you must now the documents name etc. Therefore the index as a "database" seemed to be best solution.
The problem is not just qthelp specific, its for all builders.

Is there any chance to extend the index directive with additional options?
eg:
```
.. index:: python
   :id: ID_PYTHON_REF
```
or something similar? Can you give me a hint for a possible solution i would try to implement?

**Comment by tk0miya** [comment]
+1; reasonable. But I think `:name:` is better. It is commonly used in docutils' convensions.

**Comment by kreuzberger** [comment]
Hi! Stumpled over it again (project continues). I saw that the :name: option is added to index. The resulting keyword is generated with an anchor now given via the :name: option. This is quite good. The id attribute in the keyword tag is NOT generated from the name. Why? Would this be integrated if i would make a pull request?
i expect somethink like that: `<keyword name="Mein Text" id="MY_ID" ref="File.html#MY_ID">`
If the name option is not given, the tag could be generated like bevore: `<keyword name="Mein Text" ref="File.html#index-0">`

**Pull request closed without merging**